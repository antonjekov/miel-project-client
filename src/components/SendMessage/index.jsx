import { Form, Button, Spinner, Col, Alert } from 'react-bootstrap';
import styles from './index.module.css';
import React, { useState } from 'react';
import messageSchema from "../../schemas/messageSchema";
import { useFormik } from "formik";
import messageService from "../../services/message_service";
function SendMessage(props) {

    const [showUnsuccess, setShowUnsuccess] = useState(false);
    const [showSuccess, setShowSuccess] = useState(false);
    const [isLoading, setIsLoading] = useState(false);
    //Formik integration hook
    const { handleSubmit, handleChange, errors, values, resetForm } = useFormik({
        initialValues: {
            name: '',
            email: '',
            message: ''
        },
        validationSchema: messageSchema,
        onSubmit(values, actions) {
            setIsLoading(true);
            resetForm()
            messageService.send(values)
                .then(async res => {
                    if (!res.ok) {
                        setIsLoading(false);
                        const errorsObject = await res.json()
                        actions.setErrors(errorsObject)
                        return
                    }
                    setShowSuccess(true)
                    setTimeout(()=>{setShowSuccess(false)},3000)
                    setIsLoading(false);
                    resetForm()
                })
                .catch(err => {
                    setIsLoading(false);
                    setShowUnsuccess(true)
                    setTimeout(()=>{setShowSuccess(false)},3000)
                });

        }
    });

    const buttonDisabled = (!!Object.keys(errors).length)||(values.message==='')

    return (
        <Form onSubmit={handleSubmit}>
            <div>
                <h3>Send a message</h3>
            </div>
            <br></br>
            <Form.Row>
                <Col>
                    <Form.Group >
                        <Form.Label>Name</Form.Label>
                        <Form.Control className={styles['form-control']} type="text" placeholder="Enter your name" name='name' value={values.name} onChange={handleChange} isInvalid={!!errors.name} isValid={values.name && !errors.name} />
                        <Form.Control.Feedback type='invalid' >{errors.name}</Form.Control.Feedback>
                        <Form.Control.Feedback type='valid'>Look's good</Form.Control.Feedback>

                    </Form.Group>
                </Col>
            </Form.Row>
            <Form.Row>
                <Col>
                    <Form.Group >
                        <Form.Label>Email address</Form.Label>
                        <Form.Control className={styles['form-control']} value={values.email} onChange={handleChange} type="email" placeholder="Enter email" name='email' isInvalid={!!errors.email} isValid={values.email && !errors.email} />
                        <Form.Control.Feedback type='invalid' >{errors.email}</Form.Control.Feedback>
                        <Form.Control.Feedback type='valid'>Look's good</Form.Control.Feedback>
                    </Form.Group>
                </Col>
            </Form.Row>

            <Form.Row>
                <Col>
                    <Form.Group >
                        <Form.Label>Message</Form.Label>
                        <Form.Control as='textarea' className={styles['form-control']} value={values.message} onChange={handleChange} placeholder="Write you message here" name='message' isInvalid={!!errors.message} isValid={values.message && !errors.message} />
                        <Form.Control.Feedback type='invalid' >{errors.message}</Form.Control.Feedback>
                        <Form.Control.Feedback type='valid'>Look's good</Form.Control.Feedback>
                    </Form.Group>
                </Col>
            </Form.Row>
            <Form.Row>
            {isLoading ?
                <Spinner animation="border" variant="warning" /> :
                <Button variant="warning" type="submit" disabled={buttonDisabled}>Send</Button>}
            </Form.Row>
            
            <br></br>
            <Alert onClick={()=>{setShowUnsuccess(false)}} show={showUnsuccess} variant="warning" dismissible>
                <p>Failed to send message!</p>
            </Alert>
            <Alert onClick={()=>{setShowSuccess(false)}} show={showSuccess} variant="success" dismissible>
                <p>Thank you for the message!</p>
            </Alert>
        </Form>
    )
}

export default SendMessage