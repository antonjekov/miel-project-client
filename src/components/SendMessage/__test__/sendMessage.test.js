import React from "react";
import ReactDOM from "react-dom"
import {render, cleanup, screen, fireEvent} from "@testing-library/react"
import renderer from "react-test-renderer"
import SendMessage from "../"

afterEach(cleanup)

test("inputs update on change", () =>{
    const {queryByPlaceholderText, queryByText} = render(<SendMessage/>)
    const nameInput = queryByPlaceholderText('Enter your name')
    fireEvent.change(nameInput,{target:{value: 'Anton'}})
    expect(nameInput.value).toBe('Anton')
    const emailInput = queryByPlaceholderText('Enter email')
    fireEvent.change(emailInput,{target:{value: 'anton@yahoo.com'}})
    expect(emailInput.value).toBe('anton@yahoo.com')
  })

  it('matches snapshot', ()=>{
    const tree = renderer.create(<SendMessage/>).toJSON();
  
    expect(tree).toMatchSnapshot();
  })