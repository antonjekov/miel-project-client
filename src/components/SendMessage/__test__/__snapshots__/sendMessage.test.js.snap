// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form
  className=""
  onSubmit={[Function]}
>
  <div>
    <h3>
      Send a message
    </h3>
  </div>
  <br />
  <div
    className="form-row"
  >
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label
          className="form-label"
        >
          Name
        </label>
        <input
          className="form-control form-control"
          name="name"
          onChange={[Function]}
          placeholder="Enter your name"
          type="text"
          value=""
        />
        <div
          className="invalid-feedback"
        />
        <div
          className="valid-feedback"
        >
          Look's good
        </div>
      </div>
    </div>
  </div>
  <div
    className="form-row"
  >
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label
          className="form-label"
        >
          Email address
        </label>
        <input
          className="form-control form-control"
          name="email"
          onChange={[Function]}
          placeholder="Enter email"
          type="email"
          value=""
        />
        <div
          className="invalid-feedback"
        />
        <div
          className="valid-feedback"
        >
          Look's good
        </div>
      </div>
    </div>
  </div>
  <div
    className="form-row"
  >
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label
          className="form-label"
        >
          Message
        </label>
        <textarea
          className="form-control form-control"
          name="message"
          onChange={[Function]}
          placeholder="Write you message here"
          value=""
        />
        <div
          className="invalid-feedback"
        />
        <div
          className="valid-feedback"
        >
          Look's good
        </div>
      </div>
    </div>
  </div>
  <div
    className="form-row"
  >
    <button
      className="btn btn-warning"
      disabled={true}
      type="submit"
    >
      Send
    </button>
  </div>
  <br />
</form>
`;
